src/contracts/Qearn.h

```diff
@@ -0,0 +1,680 @@
+        {
+            
+            locals.INITIALIZE_ROUNDINFO._Epoch_Bonus_Amount += state.remain_amount;
+
+            state._InitialRoundInfo.set(qpi.epoch(), locals.INITIALIZE_ROUNDINFO);
+            state._CurrentRoundInfo.set(qpi.epoch(), locals.INITIALIZE_ROUNDINFO);
+
+            state.remain_amount = 0;
+
+        }
+	_
+
+    struct END_EPOCH_locals 
+    {
+        HistoryInfo INITIALIZE_HISTORY;
+        UserInfo INITIALIZE_USER;
+        RoundInfo INITIALIZE_ROUNDINFO;
+
+        uint64 _reward_percent;
+        uint64 _reward_amount;
+        uint64 _burn_amount;
+        uint32 locked_epoch;
+        uint32 _count_Of_last_vacancy;
+        uint32 _t;
+
+    };
+
+	END_EPOCH_WITH_LOCALS
+
+        locals._count_Of_last_vacancy = 0;
+        state._Unlocked_cnt = 0;
+        locals.locked_epoch = qpi.epoch() - 52;
+        
+        locals._burn_amount = state._CurrentRoundInfo.get(locals.locked_epoch)._Epoch_Bonus_Amount;
+        
+        locals._reward_percent = QPI::div(state._CurrentRoundInfo.get(locals.locked_epoch)._Epoch_Bonus_Amount * 10000000ULL, state._CurrentRoundInfo.get(locals.locked_epoch)._Total_Locked_Amount);
+
+        for(locals._t = 0 ; locals._t < state._Locked_Last_Index; locals._t++) 
+        {
+            if(state.Locker.get(locals._t)._Locked_Epoch == locals.locked_epoch) 
+            {
+
+                if(state.Locker.get(locals._t)._Locked_Amount == 0) 
+                { //  the case to be unlocked early
+
+                    locals._count_Of_last_vacancy++; continue;
+                    
+                }
+
+                locals._reward_amount = QPI::div(state.Locker.get(locals._t)._Locked_Amount * locals._reward_percent, 10000000ULL);
+                qpi.transfer(state.Locker.get(locals._t).ID, locals._reward_amount + state.Locker.get(locals._t)._Locked_Amount);
+
+                if(state._Unlocked_cnt < MAXIMUM_UNLOCK_HISTORY) 
+                {
+
+                    locals.INITIALIZE_HISTORY._Unlocked_ID = state.Locker.get(locals._t).ID;
+                    locals.INITIALIZE_HISTORY._Unlocked_Status = 1;
+                    locals.INITIALIZE_HISTORY._Unlocked_Amount = state.Locker.get(locals._t)._Locked_Amount;
+                    locals.INITIALIZE_HISTORY._Rewarded_Amount = locals._reward_amount;
+
+                    state.Unlocker.set(state._Unlocked_cnt, locals.INITIALIZE_HISTORY);
+
+                    state._Unlocked_cnt++;
+                }
+
+                locals.INITIALIZE_USER.ID = NULL_ID;
+                locals.INITIALIZE_USER._Locked_Amount = 0;
+                locals.INITIALIZE_USER._Locked_Epoch = 0;
+
+                state.Locker.set(locals._t, locals.INITIALIZE_USER);
+
+                locals._burn_amount -= locals._reward_amount;
+
+                locals._count_Of_last_vacancy++;
+            }
+            else locals._count_Of_last_vacancy = 0;
+        }
+
+        state._Locked_Last_Index -= locals._count_Of_last_vacancy;
```

Comment:
are you sure this is not a bug?
for example: the locker amounts have below value:
```
1,2,3,4,5,6,7,8,9,10
```
with `state._Locked_Last_Index == 10`
`5,6` unlock early then the array becomes:
```
1,2,3,4,0,0,7,8,9,10
```
according to 
```
if(state.Locker.get(locals._t)._Locked_Amount == 0) 
{ //  the case to be unlocked early
    locals._count_Of_last_vacancy++; continue;
}
```
then `locals._count_Of_last_vacancy == 2` and `state._Locked_Last_Index` will be `8`

which will implicitly remove `9,10` from the locker

Reply:
I mentioned about this problem in discord. it's not a bug. in that case, `locals._count_Of_last_vacancy` should be 0.

Reply:
This logic is changed by new commit.